### PR TITLE
screen sharing for desktop.

### PR DESCRIPTION
**Pull request by cloudwebrtc**
Currently, flutter-webrtc solves macOS/Windows screen/window enumeration and can select specific sources for sharing.
The following issues still need to be resolved

Capture screen
- [x] macOS
- [x] Windows

Capture window
- [x] macOS
- [x] Windows

Watch window changes to refresh the selection dialog, if some windows are closed, or the title changes, refresh the list.
- [x] macOS
- [x] Windows

Monitor the status of the window/screen, for example, if we close the window that is in capturing, or unplug the second display HDMI cable, we need to send a notification, and the client SDK needs to automatically close the sharing, remove the video source track.
- [x] macOS
- [x] Windows

~~Another question, should we move the [example/lib/widgets/screen_select_dialog.dart](https://github.com/livekit/client-sdk-flutter/blob/screen-sharing-for-desktop/example/lib/widgets/screen_select_dialog.dart) file inside the SDK? this In this case, we need to pass in a BuildContext to build the Widget when starting the SDK.~~

- [x] Add capture source selection dialog to SDK.

related PRs:
https://github.com/flutter-webrtc/flutter-webrtc/pull/1015
https://github.com/webrtc-sdk/libwebrtc/pull/36
https://github.com/webrtc-sdk/webrtc/pull/24

imageimageimage

### CODE REVIEW

**Comment by jdde** [comment]
Hi @cloudwebrtc when it's possible to handover a deviceId (window selection) to the ScreenShareCaptureOptions I can definitely test the implementation :) just let me know, I could directly use this branch in our develop version & test it against a production system - happy to help!

**Comment by cloudwebrtc** [comment]
@jdde I'm almost done with screen sharing for windows and mac, and we have a desktop source selection dialog, but I still need a few days to improve the code and testing, which will land on the Livekit flutter SDK soon.

https://github.com/flutter-webrtc/flutter-webrtc/pull/1015

**Comment by cloudwebrtc** [comment]
> when it's possible to handover a deviceId (window selection) to the ScreenShareCaptureOptions I can definitely test the implementation :)

The idea is to integrate the screen selection dialog directly inside the SDK, when we use `localParticipant.selectDesktopSource(BuildContext context);`
In a similar way, the SDK will ask you to choose a screen or window, and you can share your screen

**Comment by jdde** [comment]
Ah nicy!
But on our side we will definitely integrate a custom selection screen. So maybe it makes sense to split that story on your side and take an iterative approach to first adjust the api in data layer to make it possible to implement such a view and then make another change, which includes this dialog?
With that approach, we could already provide earlier feedback :)

**Comment by cloudwebrtc** [comment]
@jdde Yes, we can construct LocalVideoTrack by passing in MediaStream or VideoTrack through ScreenShareCaptureOptions.

In this case, you will be able to use
https://github.com/flutter-webrtc/flutter-webrtc/blob/feat/screen-capture-event-listener-for-win/lib/src/desktop_capturer.dart#L50
to create a custom select box. And create a MediaStream through getDiaplayMedia.

Both methods will work at the same time

**Comment by cloudwebrtc** [comment]
@davidliu @hiroshihorie @davidzhao 

I think this PR can be reviewed. CI still has an error because the git repo is referenced as a test, when flutter-webrtc is merged I will change to the version on pub.

There are also several larger PRs related to this issue: xD
https://github.com/flutter-webrtc/flutter-webrtc/pull/1015
https://github.com/webrtc-sdk/libwebrtc/pull/36
https://github.com/webrtc-sdk/webrtc/pull/24

**Comment by jdde** [comment]
@cloudwebrtc I've already tested our current feature branch for screensharing on macOS with this branch as a dependency.
We're using a custom selector and hand over the sourceId.
Works really well! didn't noticed any problems. awesome work!!!

**Comment by jdde** [comment]
@cloudwebrtc okay, maybe there's one problem - but I'm not yet sure if it's connected, I updated the dependencies again to the newest flutter-webrtc version and now I'm not getting the system popups anymore, where macOS asks for permission to access camera, microphone.
Cleaned pods, cleaned flutter, fresh install. Will investigate deeper, and make sure it's because of that dependency change.

**Comment by cloudwebrtc** [comment]
hey @jdde, when your own project depends on the latest version of flutter-webrtc or livekit-sdk, after running flutter run -d macos for the first time in macOS, you can try open `macos/Runner.xcworkspace`, and then check the following options
image

In macOS also make sure the following items are added to Info.plist
https://github.com/flutter-webrtc/flutter-webrtc#ios

```xml
<key>NSCameraUsageDescription</key>
<string>$(PRODUCT_NAME) Camera Usage!</string>
<key>NSMicrophoneUsageDescription</key>
<string>$(PRODUCT_NAME) Microphone Usage!</string>
```

**Comment by jdde** [comment]
> 

Yess, I know, it's already configured in our project since months, and it's also already build with those permissions. All entries are there (also for debug & production targets), but just since this morning, macOS is not opening the popup anymore (in debug).

Anyway, looks like the issue is not connected to this pull-request :)

**Comment by cloudwebrtc** [comment]
yeah, I have also encountered a similar situation, it is easy to appear in the development environment, or m1 mac, you can try the following sqlite command to directly replace or add permission info.

```bash
sudo sqlite3 ~/Library/Application\ Support/com.apple.TCC/TCC.db "INSERT or REPLACE INTO access VALUES('kTCCServiceMicrophone','com.your.app.bundle.id',0,0,4,1,NULL,NULL,0,'UNUSED',NULL,0,1622199671);"
sudo sqlite3 ~/Library/Application\ Support/com.apple.TCC/TCC.db "INSERT or REPLACE INTO access VALUES('kTCCServiceCamera','com.your.app.bundle.id',0,0,4,1,NULL,NULL,0,'UNUSED',NULL,0,1622199671);"
sudo sqlite3 ~/Library/Application\ Support/com.apple.TCC/TCC.db "INSERT or REPLACE INTO access VALUES('kTCCServiceScreenCapture','com.your.app.bundle.id',0,0,4,1,NULL,NULL,0,'UNUSED',NULL,0,1622199671);"
```
and replace `com.your.app.bundle.id` with your app's id

**Comment by jdde** [comment]
Awesome! Thanks a lot for this hint! Will directly integrate it in our documentation. Never had this case before oO And yes, I'm using M1 - obviously a problem with macOS.

**Review by davidzhao — APPROVED** [review]
lgtm!

File: lib/src/track/options.dart [hunk 1]
```diff
@@ -77,23 +76,37 @@ class CameraCaptureOptions extends VideoCaptureOptions {
 /// Options used when creating a [LocalVideoTrack] that captures the screen.
 class ScreenShareCaptureOptions extends VideoCaptureOptions {
   final bool useiOSBroadcastExtension;
+  final String? sourceId;
+  final double? frameRate;
 
   const ScreenShareCaptureOptions({
     this.useiOSBroadcastExtension = false,
+    this.sourceId,
+    this.frameRate = 30.0,
```
[the remark this answers is not in the document]
Comment (cloudwebrtc) [reply lib/src/track/options.dart:85]: Maybe we should add `deviceId` and `maxFrameRate` to VideoCaptureOptions?

https://github.com/livekit/client-sdk-flutter/blob/main/lib/src/track/options.dart#L113

[the remark this answers is not in the document]
Comment (cloudwebrtc) [reply lib/src/track/options.dart:85]: This way camre capture can also select the device and set the frame rate

[the remark this answers is not in the document]
Comment (davidzhao) [reply lib/src/track/options.dart:85]: that's a good idea. it seem these should be applicable across the board here. what do you think @hiroshihorie ?

[the remark this answers is not in the document]
Comment (hiroshihorie) [reply lib/src/track/options.dart:85]: sounds good @cloudwebrtc

**Review by hiroshihorie — APPROVED** [review]

**Pull request merged**